x/farming/spec/03_state_transitions.md

```diff
@@ -22,55 +22,39 @@ const (
-    - ![](https://latex.codecogs.com/svg.latex?\Large&space;TR_i) : total reward amount of the staking coin for epoch i
-- Accumulated rewards from any staking position can be calculated from AUR and the staking amount of the position as below
-  - ![](https://latex.codecogs.com/svg.latex?\Large&space;x*\(\sum_{i=0}^{now}\frac{TR_i}{TS_i}-\sum_{i=0}^{start}\frac{TR_i}{TS_i}\))
-    - assuming constant staking amount for the staking epochs
-    - ![](https://latex.codecogs.com/svg.latex?\Large&space;x) : staking amount for the staking period
\ No newline at end of file
+## Stake
+
+When a farmer stakes an amount of coins, the following state transitions occur:
+
+- it reserves the amount of coins to the staking reserve pool account `StakingReservePoolAcc` 
+- it creates `QueuedStaking` object and stores the staking coins in `QueueStaking`, which are waiting in a queue until the end of epoch to move to `Staking` object
+- it imposes more gas if the farmer already has `Staking` with the same coin denom(see [07_params.md](07_params.md#DelayedStakingGasFee) for details)
+
+## Unstake
+
+When a farmer unstakes an amount of coins, the following state transitions occur:
+
+- it adds `Staking` and `QueueStaking` amounts to see if the unstaking amount is sufficient
+- it automatically withdraws rewards for the coin denom which are accumulated over the last epochs
+- it subtracts the unstaking amount of coins from `QueueStaking` first and if it is not sufficient then it subtracts from `Staking`
+- it releases the unstaking amount of coins to the farmer
+
+## Harvest (Reward Withdrawal)
+
+- it calculates `CumulativeUnitRewards` in `HistoricalRewards` object in order to get the rewards for the staking coin denom which are accumulated over the last epochs for the farmer
+- it releases the accumulated rewards to the farmer if it is not zero and decreases the `OutstandingRewards`
+- it sets `StartingEpoch` in `Staking` object
+
+## Reward Allocation
+
+Each abci end block call, the operations to update rewards allocation are to execute:
+
+++ https://github.com/tendermint/farming/blob/69db071ce30b99617b8ba9bb6efac76e74cd100b/x/farming/keeper/reward.go#L363-L426
+
+- it calculates rewards allocation information for the end of the current epoch depending on plan type `FixedAmountPlan` or `RatioPlan`
+- it distributes total allocated coins from each plan’s farming pool address `FarmingPoolAddress` to the rewards reserve pool account `RewardsReserveAcc`
+- it calculates staking coin weight for each denom in each plan and gets the unit rewards by denom
+- it updates `HistoricalRewards` and `CurrentEpoch` based on the allocation information
+- it automatically withdraws the accumulated rewards to the farmer with the given staking coin denom if `Staking` position exists
+- it deletes `QueueStaking` object after moving `QueueCoins` to `StakedCoins` in `Staking` object
+- it increases `TotalStakings` for the staking coin denom
```

Comment:
Also, these are not related with reward allocation, instead, these are related with `ProcessQueuedCoins` method.
I think we need a separate section for these end of epoch events.